### PR TITLE
KAFKA-10062: Add a method to retrieve the current timestamp as known by the Streams app

### PR DESCRIPTION
**Pull request by rohitrmd**
KIP 622: https://cwiki.apache.org/confluence/display/KAFKA/KIP-622%3A+Add+currentSystemTimeMs+and+currentStreamTimeMs+to+ProcessorContext

jira: https://issues.apache.org/jira/browse/KAFKA-10062

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by rohitrmd** [comment]
@mjsax @vvcephei @cadonna  StreamTaskTest is failing because of following issue:
StreamTask -> process(wallClockTime) method updates processor context's system time in this method: https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java#L685
AbstractProcessorContect has another systemTime field: https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractProcessorContext.java#L48 which is updated in streamTask's process method.

Now the change in pr fetches system time from StreamTask's time field: 
https://github.com/apache/kafka/pull/9744/files#diff-a76674468cda8772230fb8411717cf9068b1a363a792f32c602fb2ec5ba9efd7R1120
This time field is not updated during [process](https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java#L665) call.  Hence tests fail. 

What do you recommend in this case? Get rid of [currentSystemTimeMs](https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractProcessorContext.java#L48) from AbstractProcessorContext? If we get rid of this field, then how do we update StreamTask's [time](https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java#L73) field as it's immutable.

**Comment by rohitrmd** [comment]
@mjsax made following changes to the pr:

1. Got rid of [currentSystemTimeMs](https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractProcessorContext.java#L48) from AbstractProcessorContext, that in effect required changing other [GloblaProcessorContextImpl](https://github.com/apache/kafka/pull/9744/files#diff-66f1420c00b334fbe657e3a2c6fb5f8bcd60407dd7f87d9ed6bb1365958e01e4R121) to return current system time.  This was needed because of [this](https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/processor/internals/SourceNode.java#L87). MockProcessorContext and MockInternalProcessorContext also required similar changes. 
2. In StreamTask, had to add the method to set system [time](https://github.com/apache/kafka/pull/9744/files#diff-a76674468cda8772230fb8411717cf9068b1a363a792f32c602fb2ec5ba9efd7R1127) for test purposes only. As we are getting wall clock time from StreamTask now instead of ProcessorContext, this change was needed for these [tests](https://github.com/apache/kafka/pull/9744/files#diff-d2d5426c1d8b525415735ad409e82d4ff34ea3712a8217e4a99b421e470d4c19R497-R547).

File: streams/src/main/java/org/apache/kafka/streams/processor/ProcessorContext.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/ProcessorContext.java]: nit: `Return` without `s` -- we use imperative to write JavaDocs.

I would remove `cached` and add a dedicated second sentence about it. Also `wall-clock time` and `system time` are synonymous and thus `wall-clock system time` sounds a little odd.

What about:
```
Return the current system timestamp (also called wall-clock time) in milliseconds.
<p>
Note: this method returns the internally cached system timestamp from the Kafka Stream runtime. Thus, it may return a different value compared to `System.currentTimeMillis()`.
<p>
For a global processor, Kafka Streams does not cache system time and thus calling this method will return the same value as `System.currentTimeMillis()`.

@return the current system timestamp in milliseconds
```

File: streams/src/main/java/org/apache/kafka/streams/processor/ProcessorContext.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/ProcessorContext.java]: What about:
```
Return the current stream-time in milliseconds.
<p>
Stream-time is the maximum observed {@link TimestampExtractor record timestamp} so far
(including the currently processed record), i.e., it can be considered a high-watermark.
Stream-time is tracked on a per-task basis and is preserved across restarts and during task migration.
<p>
Note: this method is not supported for global processors (cf. {@link Topology#addGlobalStore(...)}
and {@link StreamsBuilder#addGlobalStore(...)}, because there is no concept of stream-time for this case.
Calling this method in a global processor with result in an {@link UnsupportedOperationException}.

@return the current stream-time in milliseconds
```

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ForwardingDisabledProcessorContext.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ForwardingDisabledProcessorContext.java]: Why do we throw here? It seems to be safe to get the time from `delegate` object? We only disable `forward` but nothing else. (same below)

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalProcessorContextImpl.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalProcessorContextImpl.java]: This makes sense, but we might want to document this somewhere else, ie, in the corresponding docs. To be fair, not sure atm, if we have much content about it and/or where it add it?

We should for sure document it in the JavaDocs, ie, the `ProcessorContext` interface. (cf my comment above)

About the error message:
```
throw new UnsupportedOperationException("There is no concept of stream-time for a global processor.");
```

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalProcessorContextImpl.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalProcessorContextImpl.java]: We should pass in a `Time` reference to the `GlobalProcessorContextImpl` and call `time.milliseconds()` here -- otherwise, we cannot mock the time (cf. `TopologyTestDriver`) for testing

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreToProcessorContextAdapter.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreToProcessorContextAdapter.java]: We should align the error message to the ones from above: `"StateStores can't access system time."` (same below).

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: Why is this `public` now? It seems it's only newly called in `ProcessorContextImpl` that is in the same package and thus it can stay (default) package-private.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: As above: seems this can be package-private?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: Yes, `Time` is always wall-clock time -- comment seems unnecessary

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: I think we should keep this as `final` (compare my other comments).

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: I don't think we need this method (cf. my other comment in the test code where it is used).

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/AbstractProcessorContextTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/AbstractProcessorContextTest.java]: This test seems not to be useful, as it tests the `TestProcessorContext` from below.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorContextImplTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorContextImplTest.java]: This should be two tests, one for each method.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StoreToProcessorContextAdapterTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StoreToProcessorContextAdapterTest.java:34]: Thanks for adding a test!!!

Would you mind to add tests for all other methods of `StoreToProcessorContextAdapter`, too, to get full test coverage as a side improvement?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java]: We have already a global `MockTime` object in this test that we should reused instead of creating a new one. This way, we can avoid to add `setTime` and keep the `Time` member variable `final`. -- Or is there a reason why we cannot reuse the global `MockTime` object?

Also, I don't really see any call to the newly added methods in this test, so why do we change this test method at all?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java]: Why do we not keep `task.process(10L);` ? Similar below.

File: streams/src/test/java/org/apache/kafka/test/InternalMockProcessorContext.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/test/InternalMockProcessorContext.java]: nit: revert insert of additional space

File: streams/src/test/java/org/apache/kafka/test/InternalMockProcessorContext.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/test/InternalMockProcessorContext.java]: We should pass in a `Time` parameter into the constructor of `InternalMockProcessorContext` and return `time.milliseconds()` here so we can mock the time.

If this method is unused, it's also ok to just throw an exception for now and we add mocking capabilities later when needed.

(Same below.)

File: streams/src/test/java/org/apache/kafka/test/MockInternalProcessorContext.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/test/MockInternalProcessorContext.java]: The purpose of this class seems to be to allow mocking the time -- given that we add proper time support now, I am wondering if we can actually delete the whole class?

File: streams/src/test/java/org/apache/kafka/test/NoOpProcessorContext.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/test/NoOpProcessorContext.java]: Nit: I guess we could add support if needed. Maybe better say `"Not implemented yet."` ? (Same below.)

File: streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java
Comment (mjsax) [inline streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java]: I don't think we need this change. The new test you added below, seem to re-test `ProcessorContextImpl` that is already tested in `ProcessorContextImplTest` and thus we don't need those.

File: streams/test-utils/src/main/java/org/apache/kafka/streams/processor/MockProcessorContext.java
Comment (mjsax) [inline streams/test-utils/src/main/java/org/apache/kafka/streams/processor/MockProcessorContext.java]: `timestamp` is the current record timestamp and thus should not be returned here (same below).

I seems, we need to actually extend this test-util class, what is a small change to the KIP. Note, that test-utils are public API and thus those changes also need a KIP. You don't need to worry about your KIP, we can just update it accordingly.

Seems, we need to have two more internal fields, one for each time. The public API change would be to add corresponding setters. We might also rename `setTimestamp` to `setRecordTimestamp` for clarity, we we would have three setters for record, system, and stream time.

File: streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java
Comment (mjsax) [inline streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java]: I think we can remove those two new tests.

If you really want to test it, you should rather add a `Processor` to the Topology that gets the current system and stream time and verify it this way, ie, we should rather add an integration test than a unit test.

**Review by mjsax — COMMENTED** [review]
Thanks for the PR. Made an initial pass.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java
Comment (rohitrmd) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java]: @mjsax The reason I had to change StreamTaskTest is because of following issue:
StreamTask -> process(wallClockTime) method updates processor context's system time in this method: https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java#L685
AbstractProcessorContect has another systemTime field: https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractProcessorContext.java#L48 which is updated in streamTask's process method. I removed this field from AbstractProcessorContect as we want to fetch the time from StreamTask. But the latency is measured [here](https://github.com/apache/kafka/blob/b9dfc196aaee91f4fa23fe6563261c6f2d01ac33/streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextImpl.java#L284) which is streamTask's time filed now which is not updated.
I couldn't find any other way to change time field in StreamTask unless I add setTime method in it. If i don't change time in StreamTask object, subsequent metric tests fail as stream-time is not updated. 
Do you have recommendation about how i can fix it without adding setTime method?

Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java]: In L159, we define a global mock time object:
```
private MockTime time = new MockTime();
```

that is passed into `StreamTask`. You can call `time.sleep(...)` to advance the mock time.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextUtils.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextUtils.java]: nit (not introduced in this PR, but would be a nice side cleanup): rename to `currentSystemTime(...)`

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractProcessorContext.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractProcessorContext.java:48]: @guozhangwang -- I am just wondering if removing this variable (what I think conceptually makes sense) might have an perf impact as it seems `currentSystemTimeMs` acts as some kind of cache -- and instead of using the cached time, we would call `time.currentTimeMillis()` in `ProcessorContextImpl#forwardInternal()` now.

Thoughts?

Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractProcessorContext.java:48]: I think its rationale is actually to "not advance" the system time while in the middle of processing a record, i.e. the `system time` would just be the time when we start processing / punctuating, and it would not change throughout the process / punctuate.

File: streams/src/test/java/org/apache/kafka/test/MockInternalProcessorContext.java
Comment (rohitrmd) [reply streams/src/test/java/org/apache/kafka/test/MockInternalProcessorContext.java]: @mjsax should i try to do this in another pr?

**Comment by rohitrmd** [comment]
@mjsax made changes as per the review. Can you please re-review?

File: streams/src/main/java/org/apache/kafka/streams/processor/ProcessorContext.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/ProcessorContext.java]: typo: `with` -> `will`

File: streams/src/main/java/org/apache/kafka/streams/processor/ProcessorContext.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/ProcessorContext.java]: nit:
```
`System.currentTimeMillis()` -> {@code System.currentTimeMillis()} 

or: {link System#currentTimeMillis()}
```

File: streams/src/main/java/org/apache/kafka/streams/processor/ProcessorContext.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/ProcessorContext.java]: as above

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractProcessorContext.java
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractProcessorContext.java:48]: @rohitrmd did you see Guozhang comment? Seems we should not remove this variable (what is a little unfortunate, but I guess desirable behavior) -- maybe we can rename it to `cachedSystemTimeMs`?

I guess we want the caching for the `GlobalProcessorContext`, too? (For this case, we would need to update the JavaDocs of the `ProcessorContext` interface accordingly.)

We should also add a test for the corresponding contest-test classes that they don't advance system time automatically, but only in the advance is triggered explicitly.

File: streams/src/test/java/org/apache/kafka/test/MockInternalProcessorContext.java
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/test/MockInternalProcessorContext.java]: I guess a follow up PR works, too.

**Review by mjsax — COMMENTED** [review]
Thanks for updating the PR! -- Following Guozhang's comment, I guess we should keep the system-time caching in the abstract context. Sorry for the forth-and-back...

Btw: feature freeze is 3rd March, and I would like to get this into 2.8.0 release -- hope we can push it over the finish line on time. (cf https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=173081737)

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractProcessorContext.java
Comment (rohitrmd) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractProcessorContext.java:48]: @mjsax do you recommend adding method to AbstractProcessorContext to set cachedSystemTimeMs (setCachedSystemTimeMs) or you want me to add setSystemTimeMs back to [InternalProcessorContext](https://github.com/apache/kafka/pull/9744/files#diff-34daeb287c7e79c8ccd757daa4e17d6ab585d54844f6e5e8676853762a08cedcL49) and set the system time like it was done before?

Comment (rohitrmd) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractProcessorContext.java:48]: @mjsax also I am confused about what time do we want to return when currentSystemTimeMs() is called from the ProcessorContext hierarchy. If it's called from AbstractProcessorContext object reference, should we return cachedSystemTimeMs, and if it's called from ProcessorContextImpl object reference, should we return streamTask.currentSystemTimeMs()? Or we want completely new method to return the cached time when ProcessorContextImpl#forwardInternal() is called?

Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractProcessorContext.java:48]: If there is an advantage to add the method to the `InternalProcessorContext ` interface, sure why not. Don't think it's a big difference, and because it all internal code, we can change it anytime in the future.

I think we should always return the cached time.

**Comment by mjsax** [comment]
@rohitrmd -- Seems we have till Friday to get this PR merged. Otherwise, it would slip the 2.8.0 release.

Can you address the last comments and rebase the PR to resolve merge conflicts? Or do you have any pending questions?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractProcessorContext.java
Comment (rohitrmd) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractProcessorContext.java:48]: @mjsax can you please explain again what is expected now as for me this is contrary to initial KIP. What I have understood from KIP, we wanted to return system time from Stream Task. Considering changes in this pr, 

1. We added one time field in InternalMockProcessorContext which we return when currentSystemTime() is called. 
2. When ProcessorContextImpl's currentSystemTime() method is called, we return time from streamTask's time field. 
3. We also added time field in GlobalProcessorContextImpl which we return from currentSystemTime().
4. If we add new cachedSystemTimeMs field in AbstractProcessorContext, when do you want to return this field? Are earlier changes not valid to return time from StreamTask?

Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractProcessorContext.java:48]: The main motivation to add `ProcessorContext#currentSystemTime()` was to be able to return the mocked wall-clock time in `TopologyTestDriver`. Even if we return the cached time from `AbstractProcessorContext`, we will be able to return the mocked time, as we update the cached time based on the mocked time in `TopologyTestDriver`.

- `InternalMockProcessorContext` is just for our own unit testing -- it's fine to add the new `Time` field, it's not a public facing change anyway
- Originally we changed `ProcessorContextImpl` because we remove `AbstractProcessorContext#currentSystemTime()` -- as suggested by Guozhang, we should keep the cached time in `AbstractProcessorContext()`, and thus we don't need `ProcessorContextImpl#currentSystemTime()` any longer.
- `GlobalProcessorContextImpl` is a different code path, and thus the changes of this PR are fine

> If we add new cachedSystemTimeMs field in AbstractProcessorContext, when do you want to return this field?

Yes, we want to return this field. (We get this behavior by adding back `AbstractProcessorContext#currentSystemTime()` (and the cached time in this class) and removing `ProcessorContextImpl#currentSystemTime()`.

> Are earlier changes not valid to return time from StreamTask?

Yes, I think we can revert all changes from `StreamTask`.

Does this make sense?

Comment (rohitrmd) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractProcessorContext.java:48]: @mjsax yes, thank you for the explanation. made changes as per review.

**Comment by rohitrmd** [comment]
@mjsax made changes as per the review and resolved merge conflicts.

**Comment by rohitrmd** [comment]
@mjsax 2 1. tests failing are not related to my changes. not sure how to fix them. 2. Are there kip updates i need to do?

**Comment by rohitrmd** [comment]
@mjsax have i missed 228 deadline for this pr?

**Comment by vvcephei** [comment]
Hi @rohitrmd ,

I was coincidentally just reviewing the list of KIPs that made the cutoff on Feb 3rd. Sadly, yes, this feature will not make 2.8, but we should push forward and get this merged into trunk. FYI, I'll go ahead and drop it from the 2.8 release plan and update the KIP to target 3.0.

FWIW, it's not uncommon to see some unrelated test failures. It'll be up to @mjsax to go ahead and merge or re-run the tests, based on his best judgement.

Thanks for this feature!
-John

File: streams/src/main/java/org/apache/kafka/streams/processor/ProcessorContext.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/ProcessorContext.java]: nit. can be removed

File: streams/test-utils/src/main/java/org/apache/kafka/streams/processor/MockProcessorContext.java
Comment (mjsax) [inline streams/test-utils/src/main/java/org/apache/kafka/streams/processor/MockProcessorContext.java:399]: This class is public API, so we cannot remove `setTimestamp` but can only deprecate it.

We also need to update the KIP to mention the deprecation and the newly added methods.

**Review by mjsax — APPROVED** [review]
Sorry for the delay, but after we missed the feature freeze deadline, I de-prioritized this PR a little bit.

The latest changes LGTM and this PR is ready for merging.

We only need to update the KIP (cf the latest comment). After you updated the KIP wiki page, please send a follow up email to the VOTING thread to point out the changes. I don't think we need to re-vote, but we need to give the opportunity to object (I expect that nobody will object).

If nobody objects with a few days after you sent the email, I can merge the PR.

Thanks a lot for contributing and being patience -- this PR turned out more complex that anticipated. Let us know if you have any question.

File: streams/test-utils/src/main/java/org/apache/kafka/streams/processor/MockProcessorContext.java
Comment (rohitrmd) [reply streams/test-utils/src/main/java/org/apache/kafka/streams/processor/MockProcessorContext.java:399]: @mjsax added the method setTimestamp back which sets record timestamp and added Deprecated annotation. Also updated the [kip](https://cwiki.apache.org/confluence/display/KAFKA/KIP-622%3A+Add+currentSystemTimeMs+and+currentStreamTimeMs+to+ProcessorContext).

File: streams/src/main/java/org/apache/kafka/streams/processor/ProcessorContext.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/ProcessorContext.java]:
```suggestion
     * Thus, it may return a different value compared to {@code System.currentTimeMillis()}.
```

File: streams/src/main/java/org/apache/kafka/streams/processor/ProcessorContext.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/ProcessorContext.java]:
```suggestion
```

File: streams/test-utils/src/main/java/org/apache/kafka/streams/processor/MockProcessorContext.java
Comment (mjsax) [inline streams/test-utils/src/main/java/org/apache/kafka/streams/processor/MockProcessorContext.java]:
```suggestion
     * @deprecated Since 3.0.0; use {@link MockProcessorContext#setRecordTimestamp(long)} instead.
```

**Comment by mjsax** [comment]
Merged to `trunk`.

Congrats @rohitrmd and thank you for your contribution!

**Comment by rohitrmd** [comment]
thank you @mjsax for all the help and guidance.

**Pull request merged**
